### PR TITLE
Fix multiindex loc nan

### LINKED ISSUE #43814: Using `.loc` with MultiIndex containing `np.nan` unexpected behavior
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

df = pd.DataFrame(
    {
        "temp_playlist": [0 for _ in range(15)],
        "objId": ["o1", "o2", "o3", np.nan, "o2", "o3", "o1", np.nan, "o4", "o4", "o5", "o6", "o4", "o5", "o6"],
        "x": [x for x in range(15)],
    }
)

agg_df = df.groupby(by=['temp_playlist', 'objId'], dropna=False)["x"].agg(list)
print(agg_df)
Out
temp_playlist  objId
0              o1           [0, 6]
               o2           [1, 4]
               o3           [2, 5]
               o4       [8, 9, 12]
               o5         [10, 13]
               o6         [11, 14]
               NaN          [3, 7]
Name: x, dtype: object

print(agg_df.index)
Out
MultiIndex([(0, 'o1'),
            (0, 'o2'),
            (0, 'o3'),
            (0, 'o4'),
            (0, 'o5'),
            (0, 'o6'),
            (0,  nan)], # notice this index
           names=['temp_playlist', 'objId'])

print(agg_df.loc[agg_df.index[0]])  # ok
Out
[0, 6]

# print(agg_info_df.loc[agg_info_df.index[-1]])  # KeyError: because it is (0, np.nan), wanted to get [3, 7]
print(agg_df.loc[[agg_df.index[-1]]][0].values[0])  # works with [(0, np.nan)]
Out
[3, 7]
```

### Issue Description

This issue is a follow-up of the discussion in this [SO question](https://stackoverflow.com/questions/69373456/get-values-from-dataframe-with-multiindex-index-containg-nans?noredirect=1#comment122618699_69373456).
It appears to be a bug, but if not, meaning, if this is desired behavior, it should be documented.
As shown in the `Reproducible Example`, after grouping `x` data on the `temp_playlist` and `objId` columns, there is a MultiIndex `(0, nan)`. This index is meaningful and I wanted to access the data via it as I can perform with any other index from `agg_df.index` as `agg_df.loc[<index_pos>]`. This is not possible for the index containing the `nan` (`agg_info_df.loc[agg_info_df.index[-1]]`). However, it does work if that same index is provided in a list of indices. So this seems at least inconsistent if not a bug entirely.
For more info, please consult the SO question, especially [this answer](https://stackoverflow.com/a/69375568/14094460).

### Expected Behavior

`agg_info_df.loc[(0, np.nan)]` should return `[3, 7]`

### Installed Versions

python 3.8.5, pandas 1.3.1, numpy 1.20.3

### PR DESCRIPTION
**Pull request by CloseChoice**
- [x] closes #43814 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note. 1.3.x in indexing; let's try to backport.

**Comment by jreback** [comment]
cc @phofl if you can have a look

**Review by phofl — CHANGES_REQUESTED** [review]
This is not a loc bug. The MultiIndex coming from the groupby operation is corrupt. It has the codes:

```
[[0, 0], [0, 1]]
```

This should be

```
[[0, 0], [0, -1]]
```

The nans are not identified as missing values, hence the incorrect behavior in subsequent operations. We have to fix this in the groupby. I think this is coming from the na_sentinel change in ``factorize``

**Comment by CloseChoice** [comment]
@pre-commit

**Comment by jreback** [comment]
this is WAY more complicated that before. Can you show show the groupby example and what you get now and what your expected is?

**Comment by CloseChoice** [comment]
> #36060 is related

this PR also resolves #36060. I can tests for that if that is wanted.

**Comment by jreback** [comment]
yes pls

**Comment by CloseChoice** [comment]
> yes pls

added the tests. Please note one thing:

```python
>>> import pandas as pd
>>> import numpy as np
>>> df = pd.DataFrame(
...     {
...         "animal": ["Falcon", "Falcon", "Parrot", "Parrot"],
...         "type": [np.nan, np.nan, np.nan, np.nan],
...         "speed": [380.0, 370.0, 24.0, 26.0],
...     }
... )
>>> speed = df.groupby(["animal", "type"], dropna=False)["speed"].first()
>>> # Reconstruct same index to allow for multiplication.
>>> ix_wing = pd.MultiIndex.from_tuples(
...     [("Falcon", np.nan), ("Parrot", np.nan)], names=["animal", "type"]
... )
>>> wing = pd.Series([42, 44], index=ix_wing)
>>> 
>>> result = wing * speed
>>> wing.index.dtypes
animal    object
type      object
dtype: object
>>> speed.index.dtypes
animal     object
type      float64
dtype: object

# dtype of multiindex is dependent of the order of factors
>>> (wing * speed).index.dtypes
animal    object
type      object
dtype: object

>>> (speed * wing).index.dtypes
animal     object
type      float64
dtype: object
```

Therefore I only multiply `wing * speed` and not `speed * wing`. I guess this is a bug.

**Comment by CloseChoice** [comment]
@github-actions pre-commit

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -859,7 +859,21 @@ def ngroups(self) -> int:
     def reconstructed_codes(self) -> list[np.ndarray]:
         codes = self.codes
         ids, obs_ids, _ = self.group_info
-        return decons_obs_group_ids(ids, obs_ids, self.shape, codes, xnull=True)
+        reconstructed_codes = decons_obs_group_ids(
```
Comment (jreback) [inline pandas/core/groupby/ops.py:862]: umm would be better to actually do this in `decons_obs_group_ids`

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 2]
```diff
@@ -370,3 +370,51 @@ def test_groupby_nan_included():
...
+            "x": [1, 2, 3, 4],
+        }
+    )
+
+    grouped_df = df.groupby(by=["temp_playlist", "objId"], dropna=False)["x"].sum()
+    expected = pd.MultiIndex.from_arrays(
+        [[0, 0], ["o1", na]], names=["temp_playlist", "objId"]
+    )
+    result = grouped_df.index
+    assert all((res == ex).all() for res, ex in zip(result.codes, expected.codes))
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby_dropna.py:394]: comes the actual result with an expected frame and use tm.assert_frame_equals

File: pandas/core/groupby/grouper.py [hunk 3]
```diff
@@ -690,6 +693,13 @@ def _codes_and_uniques(self) -> tuple[np.ndarray, ArrayLike]:
             codes, uniques = algorithms.factorize(
                 self.grouping_vector, sort=self._sort, na_sentinel=na_sentinel
             )
+            # GH43943, store placeholder for (np.nan, pd.NaT, np.datetime64("NaT", "ns"),
+            # np.timedelta64("NaT", "ns")), will later be replaced by -1 in
+            # pandas/core/groupby:reconstructed_codes
+            if not self._dropna:
+                if isna(self.grouping_vector).any():
+                    self._na_placeholder = max(codes)
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/groupby/grouper.py:701]: umm why do we need to store this? can we not just do max(codes) on reconstrutcion?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:862]: i dunno, not wild about core.sorting needing to know anything about self._groupings

File: pandas/core/groupby/grouper.py [hunk 4]
```diff
@@ -690,6 +693,13 @@ def _codes_and_uniques(self) -> tuple[np.ndarray, ArrayLike]:
             codes, uniques = algorithms.factorize(
                 self.grouping_vector, sort=self._sort, na_sentinel=na_sentinel
             )
+            # GH43943, store placeholder for (np.nan, pd.NaT, np.datetime64("NaT", "ns"),
+            # np.timedelta64("NaT", "ns")), will later be replaced by -1 in
+            # pandas/core/groupby:reconstructed_codes
+            if not self._dropna:
+                if isna(self.grouping_vector).any():
+                    self._has_na_placeholder = True
```
Comment (jbrockmendel) [inline pandas/core/groupby/grouper.py:701]: can this be a cache_readonly?  we try to avoid statefulness

**Comment by CloseChoice** [comment]
@github-actions pre-commit

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jreback) [reply pandas/core/groupby/ops.py:862]: hmm i take this back, this now makes sorting dependent on groupby.ops which is very strange.

@CloseChoice so instead am ok to pass *something* do `decons_obs_group_ids` but the something should not be a list of grouping, rather list of codes or similar

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by CloseChoice** [comment]
@jreback : am now passing only a list of levels where `na` needs to be reconstructed.

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (CloseChoice) [reply pandas/core/groupby/ops.py:868]: changed this to pass only a list where we need to reconstruct the codes for na

File: pandas/core/groupby/grouper.py [hunk 5]
```diff
@@ -34,6 +34,7 @@
     Categorical,
     ExtensionArray,
 )
+from pandas.core.base import isna
```
Comment (jbrockmendel) [inline pandas/core/groupby/grouper.py:37]: `from pandas.core.dtypes.missing import isna`

Comment (CloseChoice) [reply pandas/core/groupby/grouper.py]: done

File: doc/source/whatsnew/v1.4.0.rst [hunk 6]
```diff
@@ -784,6 +784,7 @@ Groupby/resample/rolling
 - Bug in :meth:`GroupBy.nth` failing on ``axis=1`` (:issue:`43926`)
 - Fixed bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` not respecting right bound on centered datetime-like windows, if the index contain duplicates (:issue:`3944`)
 - Bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` when using a :class:`pandas.api.indexers.BaseIndexer` subclass that returned unequal start and end arrays would segfault instead of raising a ``ValueError`` (:issue:`44470`)
+- Bug in :meth:`DataFrame.groupby` when grouping on multiple columns where at least one includes ``np.nan`` which resulted in a ``KeyError`` when the ``np.nan`` containing index was selected with :meth:`Series.loc` (:issue:`43814`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:787]: if you can move to 1.5

**Review by jreback — CHANGES_REQUESTED** [review]
this was good last time i looked, if you can merge master

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears to have gone stale. If interested in continuing please merge the main branch and move the whatsnew note to 1.5, and we can reopen.

**Pull request closed without merging**